### DISCUSSION TITLE
Feature Sliced review

### DISCUSSION THREAD

**Discussion by azinit** [Karvacky-Roma/loripsum-generator#9] [General]
Опишу здесь основные моменты, надеюсь поможет 🍰 

UPD: Буду офк накидывать чисто со стороны проектирования по FS

> PR не оч подходил, т.к. нужно еще отсылаться и на не задетые диффом файлы

**Comment by azinit** [comment]
**Начну с того, что "хорошо"))**

[`praise:`](https://conventionalcomments.org/#labels) Глобальные и инициализирующие стили - по канонам 🚀 

https://github.com/yesnoruly/loripsum-generator/blob/2d9e07ebe0a6d181d475dc1e1150e388d1fe627b/src/app/index.css#L3-L4

**Comment by azinit** [comment]
[`praise:`](https://conventionalcomments.org/#labels) Выделять сразу базовый shared UIKit - тоже топ

https://github.com/yesnoruly/loripsum-generator/blob/2d9e07ebe0a6d181d475dc1e1150e388d1fe627b/src/shared/ui/index.jsx#L1-L7

**Comment by azinit** [comment]
[`praise:`](https://conventionalcomments.org/#labels) Тож отлично разделять явно апишки ресурсов, которые используются

https://github.com/yesnoruly/loripsum-generator/blob/2d9e07ebe0a6d181d475dc1e1150e388d1fe627b/src/shared/api/index.js#L1

**Comment by azinit** [comment]
[`praise:`](https://conventionalcomments.org/#labels) Тоже неплохо обыграл View и ViewModel )

Чтоб компонент не разбухал

https://github.com/yesnoruly/loripsum-generator/blob/2d9e07ebe0a6d181d475dc1e1150e388d1fe627b/src/features/copy-to-clipboard/ui.jsx#L3-L21

**Comment by azinit** [comment]
[`praise:`](https://conventionalcomments.org/#labels) Отлично, что сразу есть понимание, что надо заводить сущность, даже если там хранится только модель (и даже если используется в одном месте)

https://github.com/yesnoruly/loripsum-generator/blob/2d9e07ebe0a6d181d475dc1e1150e388d1fe627b/src/entities/text/model.js#L1-L7

**Comment by azinit** [comment]
**Теперь к `toImprove`)**

[`issue(critical):`](https://conventionalcomments.org/#labels) Сейчас "пойдет", но позже роутинг должен явно декларироваться)

[`suggestion(critical):`](https://conventionalcomments.org/#labels) Даж если роутера нет и страница по факту одна - то желательно этот момент отдельно задекларировать. Чтоб когда появятся доп. страницы - не было больно все это мигрировать (ну и чтобы в app логика не разбухала)

Можно прям на уроне `pages/index.tsx` реэкспортить обертку, где напрямую home юзается. Но для аппа это будет обычный роутинг, который потом всегда будет легко поменять

https://github.com/yesnoruly/loripsum-generator/blob/2d9e07ebe0a6d181d475dc1e1150e388d1fe627b/src/app/index.jsx#L4-L7

**Comment by azinit** [comment]
[`issue(critical):`](https://conventionalcomments.org/#labels) Так у нас **App** или **HomePage** уровень абстракции?)

[`suggestion(critical):`](https://conventionalcomments.org/#labels) Тут лучше бы явно определить разметку и стили app в соответствующем слое, а для страниц оставить их уровень реализации

[`thoughts:`](https://conventionalcomments.org/#labels)
(не говорю вставлять сразу хедер/футер в апп - но это явно не должно описываться в странице - потом будет больно добавлять новые иначе)

https://github.com/yesnoruly/loripsum-generator/blob/2d9e07ebe0a6d181d475dc1e1150e388d1fe627b/src/pages/home/index.css#L1-L12

**Comment by azinit** [comment]
[`issue(critical):`](https://conventionalcomments.org/#labels) Можно заметить и по стилям и по разметке - что там уж слишком разбухшая логика)

[`thoughts:`](https://conventionalcomments.org/#labels) При этом тот же `CopyToClipboard` почему-то вынесен в отдельную фичу, но не остальное)

[`suggestion(critical):`](https://conventionalcomments.org/#labels) Щас не особо больно, но по-хорошему бы декомпозировать это все дело:

- `features/generate-text` - форма для генерации lorem-заглушки со всеми нужными инпутами
   Пока кажется еще сильней ее дробить не имеет смысл, разве что CopyToClipboard можно действительно отдельно подержать (хотя это возможно вообще слой shared).
   Также можно временно перенести стор связанный с текстом ближе к фиче, а там посмотреть - насколько стоит делать его глобальным (локальный стор всегда предпочтительней глобального, т.к. такое проще дебажить/развивать - хотя это мы уже в датафлоу уходим)
- `shared/ui/empty` - обычно для этого даж отдельный компонент выделяют (можно глянуть [здесь](https://ant.design/components/empty/))

[`thoughts:`](https://conventionalcomments.org/#labels) Вот. И даж при такой декомпозиции - у тебя уже фича берет на себя значительную часть ответственности. А страница по сути лишь композит и дополняет нужной логикой на своем уровне

https://github.com/yesnoruly/loripsum-generator/blob/2d9e07ebe0a6d181d475dc1e1150e388d1fe627b/src/pages/home/index.jsx#L30-L55

https://github.com/yesnoruly/loripsum-generator/blob/2d9e07ebe0a6d181d475dc1e1150e388d1fe627b/src/pages/home/index.css#L14-L37

**Reply by azinit** [reply]
@yesnoruly Кста я имел в виду чтоб само отображение текста - тоже в форме было)

Это ее зона отвественности, не айс такое на странице заново собирать каждый раз

![image](https://user-images.githubusercontent.com/42924400/141677164-bfd01588-17ff-4250-8bb1-1f7ec7cc746c.png)

**Reply by yesnoruly** [reply]
@azinit 
> ...чтобы отображение текста - тоже в форме было

В форме имеется ввиду в фиче генерации текста? Если да то как туда это поместить?
У меня есть предположение сделать таким образом:
```
features/generate-text/          # Фича
├── ui/                          
|    ├── form.jsx                # Собственно сама форма
|    └── view.jsx                # Отображение текста
├── model.js                     # Модель фичи
└── index.js
```

**P.s** на всякий случай для контекста: Я говорю о том чтобы переместить этот кусок кода во view (поправь если не то) 
![Screenshot_11](https://user-images.githubusercontent.com/64963734/141681361-43c439a7-6620-4160-82eb-928f8074ccc3.png)

**Reply by azinit** [reply]
Да я бы в фичу весь output положил

Просто с пропсом, который бы через композицию позволил бы фичу `copy-clipboard` пустить

![image](https://user-images.githubusercontent.com/42924400/141700188-87188264-39cb-4302-a9d9-9aecd305ef2b.png)

**Reply by azinit** [reply]
Внутри да, можно явно в UI разграничить, что есть **тулбар для генерации фичи** и **текст для отображения**

Поначалу их можно даже реэкспортить отдельно, но кмк - можно сразу вместе пока)

**Reply by azinit** [reply]
Для уточнения можно задать себе вспомогательные вопросы:
- "Если мне нужно будет поправить логику фичи, то где будет проходить ось изменений? (в идеале - везде около самого юзкейса, а все что идет вне - значит должно быть там по [критериям слоев, тобишь скоупам ответственности](https://feature-sliced.design/docs/get-started/cheatsheet))
- "А для самой фичи - что относится чисто к ней и обрабатываемому юзкейсу? А что должно определяться снаружи?"

**Reply by yesnoruly** [reply]
Видимо на этом моменте придётся выносить `CopyClipboard` в `shared`, чтобы с чистой душой напрямую использовать его во `view.jsx.` Как думаешь норм затея или лучше оставить пока как есть и через композицию (я не совсем понял как именно xd)

**Reply by azinit** [reply]
В целом да, можно было бы и оставить в фичах с композицией
Но тут прям как будто [все сигналы](https://github.com/yesnoruly/loripsum-generator/discussions/9#discussioncomment-1629503) на то, что это `shared/lib` скорее

**Reply by azinit** [reply]
@yesnoruly Некритично, но я бы всеж copy-clipboard и все обвязки прям в `shared/lib/clipboard` положил)

А то прям ось изменений как будто искуственно размазывается, хотя могла быть локализована
![image](https://user-images.githubusercontent.com/42924400/142255911-a6470599-1665-4c8e-83d4-3e2d0e991d09.png)

**Reply by yesnoruly** [reply]
[Fixed](https://github.com/yesnoruly/loripsum-generator/commit/1d57e6ceb56fb9730f8c3f6bd45762422ee16745)

**Reply by azinit** [reply]
Пушка бомба)

**Comment by azinit** [comment]
[`nitpic(non-critical):`](https://conventionalcomments.org/#labels) Некритично, но для консистнетности и конвенциональности - папку с иконками назвал бы `assets` 

https://github.com/yesnoruly/loripsum-generator/blob/2d9e07ebe0a6d181d475dc1e1150e388d1fe627b/src/features/copy-to-clipboard/ui.jsx#L7-L8

[`thoughts:`](https://conventionalcomments.org/#labels) Ну и да - тож надо следить за сложностью папки, чтоб было считываемо какая логика к какому сегменту относится

Пока норм, но потом по мере сложности фичи - понадобится директории выделять почти наверняка
![image](https://user-images.githubusercontent.com/42924400/141376691-b991fb12-3584-4a3e-938e-161b60513e50.png)

**Comment by azinit** [comment]
[`thoughts:`](https://conventionalcomments.org/#labels) На самом деле большой вопрос как уже говорил - является ли это фичой... Пока можно оставить, но что-то мне подсказывает, что придется в shared/lib положить, т.к. логика достаточно близкая к веб-платформе)

[`thoughts:`](https://conventionalcomments.org/#labels) Кста да - чтобы это попало в фичу генерации текста - не обязательно напрямую дергать - можно как раз на уроне страницы композить как считаешь нужным
_(а фича чтобы просто имела пропс - через который можно передать что угодно)_

https://github.com/yesnoruly/loripsum-generator/blob/2d9e07ebe0a6d181d475dc1e1150e388d1fe627b/src/features/copy-to-clipboard/ui.jsx#L2-L20

https://github.com/yesnoruly/loripsum-generator/blob/2d9e07ebe0a6d181d475dc1e1150e388d1fe627b/src/features/copy-to-clipboard/lib.js#L1-L10

**Comment by azinit** [comment]
[`nitpick(non-critical):`](https://conventionalcomments.org/#labels) Пока некритично, но прям цепляется глаз за относительные импорты слоев)

Потом это может неплохо так выстрелить в колено, поэтому для импорта **внешних ресурсов** лучше сразу использовать абсолютные импорты

> См. подробнее [здесь](https://github.com/feature-sliced/documentation/discussions/52#discussioncomment-477443)

https://github.com/yesnoruly/loripsum-generator/blob/2d9e07ebe0a6d181d475dc1e1150e388d1fe627b/src/pages/home/index.jsx#L6-L11

**Comment by azinit** [comment]
[`suggestion(non-critical):`](https://conventionalcomments.org/#labels) Пока норм, и в целом оправдано с CSSinJS что каждый компонент единым файлом, а не директорией.

Но тем не менее рано или поздно компоненты эти начнут друг друга дергать, т.к. даже UIKit бывает разной сложности

> Select,Button => ComboBox

Можно глянуть [атомик ](https://atomicdesign.bradfrost.com/table-of-contents/) и по нему потом раскладывать компоненты, чтобы они друг друга не кросс импортили жестко

![image](https://user-images.githubusercontent.com/42924400/141377229-23fb18f4-ca2b-4045-b3b3-08212fcf21f5.png)

**Comment by azinit** [comment]
**За остальное глаз не зацепился вродь, но и у проекта сложность не шибко большая на старте)**

На самом деле еще стоит понимать - что концепции эти часто мешает применять уже существующее понимание принципов (GRASP, DRY) + нехватка контекста, чтобы правильно применять концепции FS (критерии слоев)

Но уже оч круто, что есть понимание - что можно дробить на слои логику, а не держать "на подольше" рядом с местом использования. Это рано или поздно выстреливает, особенно когда проект больше 1 человека мейнтейнит)

**Comment by azinit** [comment]
Если все эти правки внести, то вполне можно добавить репе лейбл [`feature-sliced`](https://github.com/topics/feature-sliced) и можно будет к нам в [примеры FS проектов занести спокойно](https://github.com/feature-sliced/documentation/edit/master/website/src/pages/examples/_config.js), если есть желание)

**Reply by yesnoruly** [reply]
Ну вообще можно было бы добавить, но по сравнению, если так можно выразиться, с "соседними" проектами - этот выглядит пустовато.
Надо будет, наверное, ещё над каким-то функционалом подумать (аля темизация какая-нибудь или апишку с более обширными функциями подыскать)

**Reply by azinit** [reply]
ТудуАппа тож не гиперсложная там)

Но кста да - как варик добавить темизации той же можно

**Reply by azinit** [reply]
Или посмотреть что есть в похожих lorem-генераторах

**Reply by yesnoruly** [reply]
@azinit Готово, [ждёт своего ревью](https://github.com/feature-sliced/documentation/pull/364))

**Reply by azinit** [reply]
@yesnoruly Все топчик, спасибо что практически все правки внес  🚀

**Reply by azinit** [reply]
Уточнить разве что хотел бы - а тыж вроде хотел добавлять в примеры, когда добавишь доп функционал, не?)
Там темизацию/доп фичи для генератора

Или пока не особо для этого желание/время есть?)

**Reply by yesnoruly** [reply]
@azinit Да, времени особо сейчас не густо. Но я [дорожную карту проекта](https://github.com/yesnoruly/loripsum-generator#roadmap) основную наметил, думал уже в процессе всё дописывать)

**Comment by azinit** [comment]
Если останутся вопросы - пингуй)

**Comment by yesnoruly** [comment]
Так-с, ну вроде все критикал и около того моменты пофиксены)

> финальный пр смотреть тут (#11)

**Comment by azinit** [comment]
@yesnoruly Плюс кста некритично, но можно такую блямбу добавить при желании)

И спасибо что лейбл указал

![image](https://user-images.githubusercontent.com/42924400/142260205-e074d765-e708-483b-84bd-0758b3d333ee.png)


```
[![FeatureSliced](https://img.shields.io/badge/Powered%20by-%F0%9F%8D%B0%20Feature%20Sliced-%235c9cb5)](https://feature-sliced.design/)
```

**Comment by yesnoruly** [comment]
@azinit На счёт этого:
![Screenshot_3](https://user-images.githubusercontent.com/64963734/142261670-874978b4-5521-404a-a80e-1d282a26d2c6.png)
 Не надо было этого делать наверное, да?) Ну впринципе если подумать - не настолько мажорный пр, чтобы по списочку тому проходиться)

**Reply by azinit** [reply]
Ну это конвенции прост)

Не зря же написано:
![image](https://user-images.githubusercontent.com/42924400/142262614-07a53a74-cbbc-403c-9896-385d04d6b94c.png)


Даж если тебе фактически не приходилось делать, не обязательно было прям сносить нафиг чеклист)
_(хотя по сути ты как минимум пункты про "CONTRIBUTING" и "описание" выполнил)_

**Reply by yesnoruly** [reply]
Понял, буду знать)
